students/hirot_directory/session04/trigrams_hiro.py

```diff
@@ -0,0 +1,51 @@
+words = "I wish may I wish I might".split()
+
+print("The approach from class")
+
+trigrams = {}
+
+for i in range(len(words)-2):
+    pair = tuple(words[i:i+2])
+    follower = words[i+2]
+    print(pair, follower)
+    if pair in trigrams:
+        trigrams[pair].append(follower)
+    else:
+        trigrams[pair] = [follower] #key:pair
+
+print(trigrams)
+
+# .item :
+
+print("The approach with .item()")
+
+words = "I wish may I wish I might".split()
+
+trigrams = {}
+
+for i in range(len(words)-2):
+    pair = tuple(words[i:i+2])
+    follower = words[i+2]
+    print(pair, follower)
+    trigrams[pair] = follower
+
+print(trigrams.items())
+
+
+# .setdefault :
+
+print("THe approach with .setdefault")
+
+words = "I wish may I wish I might".split()
+
+trigrams = {}
+
+for i in range(len(words)-2):
+    pair = tuple(words[i:i+2])
+    follower = words[i+2]
+    print(pair, follower)
+    if pair not in trigrams:
+        trigrams.setdefault(pair, follower)        
+    # else:
+    #     trigrams[pair].append(follower)        
+print(trigrams)
```

Comment:
The purpose of setdefault is to initialize the value to something, if it is not there already, so you are close. However, you cannot append to a string (which is the error you were getting), so what you really want to do is setdefault to be a list, so you can append. So, your job is to figure out how setdefault would get rid of this conditional:
```
if pair not in trigrams:
    trigrams[pair] = [follower]
else:
    trigrams[pair].append(follower)
```
Note that is the same conditional you have in the first case, that we did in class. If you use setdefault to create a list (with nothing in it!), then you can append immediately, and get rid of your if/else statement. Here is an example: https://www.saltycrane.com/blog/2010/02/python-setdefault-example/